src/sign/keyset.rs

```diff
@@ -0,0 +1,1906 @@
+    let mut actions = Vec::new();
+    match rollstate {
+        RollState::Propagation1 => {
+            actions.push(Action::UpdateDnskeyRrset);
+            actions.push(Action::ReportDnskeyPropagated);
+        }
+        RollState::CacheExpire1(_) => (),
+        RollState::Propagation2 => {
+            actions.push(Action::UpdateRrsig);
+            actions.push(Action::ReportRrsigPropagated);
+        }
+        RollState::CacheExpire2(_) => (),
+        RollState::Done => {
+            actions.push(Action::UpdateDnskeyRrset);
+        }
+    }
+    actions
+}
+
+fn csk_roll(rollop: RollOp, ks: &mut KeySet) -> Result<(), Error> {
+    match rollop {
+        RollOp::Start(old, new) => {
+            // First check if the current CSK-roll state is idle. We need
+            // to check all conflicting key rolls as well. The way we check
+            // is to allow specified non-conflicting rolls and consider
+            // everything else as a conflict.
+            if let Some(rolltype) = ks.rollstates.keys().next() {
+                if *rolltype == RollType::CskRoll {
+                    return Err(Error::WrongStateForRollOperation);
+                } else {
+                    return Err(Error::ConflictingRollInProgress);
+                }
+            }
+            // Check if we can move the states of the keys
+            ks.update_csk(Mode::DryRun, old, new)?;
+            // Move the states of the keys
+            ks.update_csk(Mode::ForReal, old, new)
+                .expect("Should have been check with DryRun");
+        }
+        RollOp::Propagation1 => {
+            // Set the visiable time of new KSKs, ZSKs and CSKs to the current
+            // time.
+            let now = UnixTime::now();
+            for k in &mut ks.keys {
+                match &k.keytype {
+                    KeyType::Ksk(keystate)
+                    | KeyType::Zsk(keystate)
+                    | KeyType::Csk(keystate, _) => {
+                        if keystate.old || !keystate.present {
+                            continue;
+                        }
+
+                        k.timestamps.visible = Some(now.clone());
+                    }
+                    KeyType::Include(_) => (),
+                }
+            }
+        }
+        RollOp::CacheExpire1(ttl) => {
+            for k in &mut ks.keys {
+                let keystate = match &k.keytype {
+                    KeyType::Ksk(keystate)
+                    | KeyType::Zsk(keystate)
+                    | KeyType::Csk(keystate, _) => keystate,
+                    KeyType::Include(_) => continue,
+                };
+                if keystate.old || !keystate.present {
+                    continue;
+                }
+
+                let visible = k
+                    .timestamps
+                    .visible
+                    .as_ref()
+                    .expect("Should have been set in Propagation1");
```

Comment:
This can't be guaranteed, right -- shouldn't this return an `Err`?

Reply:
Yes, the implementation of the key roll algorithm has to guarantee this. It is not the user's problem. If it becomes a problem we could add a consistency check when loading a keyset.